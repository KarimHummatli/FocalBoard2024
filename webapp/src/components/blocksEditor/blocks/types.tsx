import {ContentBlockTypes} from '../../../blocks/block'

export type BlockInputProps = {
    onChange: (value: string) => void
    value: string
    onCancel: () => void
    onSave: (val: string) => void
}

export type ContentType = {
    name: ContentBlockTypes
    displayName: string
    slashCommand: string
    prefix: string
    Input: React.FunctionComponent<BlockInputProps>
    render: (value: string) => React.ReactNode
    runSlashCommand: (changeType: (contentType: ContentType) => void, changeValue: (value: string) => void, ...args: string[]) => void
    nextType?: string
}
