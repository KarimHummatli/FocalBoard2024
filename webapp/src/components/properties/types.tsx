import React from 'react'
import {IntlShape} from 'react-intl'

import {Card} from '../../blocks/card'
import {Board, IPropertyTemplate, PropertyTypeEnum} from '../../blocks/board'
import {Option} from '../calculations/options'

export type PropertyProps = {
    card: Card,
    board: Board,
    readOnly: boolean,
    propertyValue: string | string[],
    propertyTemplate: IPropertyTemplate,
    showEmptyPlaceholder: boolean,
}

export type PropertyType = {
    Editor: React.FunctionComponent<PropertyProps>,
    Value: React.FunctionComponent<PropertyProps>,
    name: string,
    type: PropertyTypeEnum,
    calculationOptions: Option[],
    displayName: (intl: IntlShape) => string,
    displayValue: (value: string | string[] | undefined, card: Card, template: IPropertyTemplate, intl: IntlShape) => string | string[] | undefined,
}

