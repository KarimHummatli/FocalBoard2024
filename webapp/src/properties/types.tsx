import React from 'react'
import {IntlShape} from 'react-intl'

import {Card} from '../blocks/card'
import {Board, IPropertyTemplate, PropertyTypeEnum} from '../blocks/board'
import {Option} from '../components/calculations/options'

export type PropertyProps = {
    card: Card,
    board: Board,
    readOnly: boolean,
    propertyValue: string | string[],
    propertyTemplate: IPropertyTemplate,
    showEmptyPlaceholder: boolean,
}

export type PropertyType = {
    Editor: React.FunctionComponent<PropertyProps>,
    name: string,
    type: PropertyTypeEnum,
    isDate?: boolean,
    canGroup?: boolean,
    canFilter?: boolean,
    isReadOnly?: boolean,
    calculationOptions: Option[],
    displayName: (intl: IntlShape) => string,
    displayValue: (value: string | string[] | undefined, card: Card, template: IPropertyTemplate, intl: IntlShape) => string | string[] | undefined,
    exportValue: (value: string | string[] | undefined, card: Card, template: IPropertyTemplate, intl: IntlShape) => string,
    valueLength: (value: string | string[] | undefined, card: Card, template: IPropertyTemplate, intl: IntlShape, fontDescriptor: string, perItemPadding: number) => number,
    getDateFrom?: (value: string | string[] | undefined, card: Card) => Date,
    getDateTo?: (value: string | string[] | undefined, card: Card) => Date,
}
