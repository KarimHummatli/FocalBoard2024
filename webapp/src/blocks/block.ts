// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {Utils} from '../utils'

const contentBlockTypes = ['text', 'image', 'divider', 'checkbox'] as const
const blockTypes = [...contentBlockTypes, 'board', 'view', 'card', 'comment', 'unknown'] as const
type ContentBlockTypes = typeof contentBlockTypes[number]
type BlockTypes = typeof blockTypes[number]

interface IBlock {
    id: string
    parentId: string
    rootId: string
    createdBy: string
    modifiedBy: string

    schema: number
    type: BlockTypes
    title: string
    fields: Record<string, any>

    createAt: number
    updateAt: number
    deleteAt: number
}

class Block implements IBlock {
    id: string
    parentId: string
    rootId: string
    createdBy: string
    modifiedBy: string

    schema: number
    type: BlockTypes
    title: string
    fields: Record<string, any>

    createAt: number
    updateAt: number
    deleteAt: number

    constructor(block?: IBlock) {
        this.id = block?.id || Utils.createGuid()
        this.schema = 1
        this.parentId = block?.parentId || ''
        this.rootId = block?.rootId || ''
        this.createdBy = block?.createdBy || ''
        this.modifiedBy = block?.modifiedBy || ''
        this.type = block?.type || 'unknown'

        // Shallow copy here. Derived classes must make deep copies of their known properties in their constructors.
        this.fields = block?.fields ? {...block?.fields} : {}

        this.title = block?.title || ''

        const now = Date.now()
        this.createAt = block?.createAt || now
        this.updateAt = block?.updateAt || now
        this.deleteAt = block?.deleteAt || 0
    }
}

export type {ContentBlockTypes, BlockTypes}
export {blockTypes, contentBlockTypes, IBlock, Block}
