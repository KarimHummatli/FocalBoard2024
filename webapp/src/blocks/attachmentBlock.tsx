// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {Block, createBlock} from './block'

type AttachmentBlockFields = {
    attachmentId: string
}

type AttachmentBlock = Block & {
    type: 'attachment'
    fields: AttachmentBlockFields
    isUploading: boolean
}

function createAttachmentBlock(block?: Block): AttachmentBlock {
    return {
        ...createBlock(block),
        type: 'attachment',
        fields: {
            attachmentId: block?.fields.attachmentId || '',
        },
        isUploading: false,
    }
}

export {AttachmentBlock, createAttachmentBlock}
