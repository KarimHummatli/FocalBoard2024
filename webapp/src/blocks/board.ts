// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {Utils} from '../utils'

import {Block, IBlock} from './block'
import {Card} from './card'

type PropertyType = 'text' | 'number' | 'select' | 'multiSelect' | 'date' | 'person' | 'file' | 'checkbox' | 'url' | 'email' | 'phone' | 'createdTime' | 'createdBy' | 'updatedTime' | 'updatedBy'

interface IPropertyOption {
    id: string
    value: string
    color: string
}

// A template for card properties attached to a board
interface IPropertyTemplate {
    id: string
    name: string
    type: PropertyType
    options: IPropertyOption[]
}

type BoardFields = {
    icon: string
    description: string
    showDescription?: boolean
    isTemplate?: boolean
    cardProperties: IPropertyTemplate[]
}

class Board extends Block {
    fields: BoardFields

    constructor(block?: IBlock) {
        super(block)
        this.type = 'board'

        let cardProperties: IPropertyTemplate[] = []
        if (block?.fields.cardProperties) {
            // Deep clone of card properties and their options
            cardProperties = block?.fields.cardProperties.map((o: IPropertyTemplate) => {
                return {
                    id: o.id,
                    name: o.name,
                    type: o.type,
                    options: o.options ? o.options.map((option) => ({...option})) : [],
                }
            })
        }

        const selectProperties = cardProperties.find((o) => o.type === 'select')
        if (!selectProperties) {
            const property: IPropertyTemplate = {
                id: Utils.createGuid(),
                name: 'Status',
                type: 'select',
                options: [],
            }
            cardProperties.push(property)
        }

        this.fields = {
            showDescription: block?.fields.showDescription || false,
            description: block?.fields.description || '',
            icon: block?.fields.icon || '',
            isTemplate: block?.fields.isTemplate || false,
            cardProperties,
        }
    }

    duplicate(): Board {
        const board = new Board(this)
        board.id = Utils.createGuid()
        return board
    }
}

type BoardGroup = {
    option: IPropertyOption
    cards: Card[]
}

export {Board, PropertyType, IPropertyOption, IPropertyTemplate, BoardGroup}
