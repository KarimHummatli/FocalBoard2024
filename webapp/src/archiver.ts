// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {IAppWindow} from './types'
import {ArchiveHeader, ArchiveLine, BlockArchiveLine} from './blocks/archive'
import {Block} from './blocks/block'
import {Board} from './blocks/board'
import {LineReader} from './lineReader'
import mutator from './mutator'
import {Utils} from './utils'

declare let window: IAppWindow

class Archiver {
    static async exportBoardArchive(board: Board): Promise<void> {
        this.exportArchive(mutator.exportArchive(board.id))
    }

    static async exportFullArchive(): Promise<void> {
        this.exportArchive(mutator.exportArchive())
    }

    private static exportArchive(prom: Promise<Response>): void {
        // TODO:  don't download whole archive before presenting SaveAs dialog.
        prom.then((response) => {
            response.blob().
                then((blob) => {
                    const link = document.createElement('a')
                    link.style.display = 'none'

                    const date = new Date()
                    const filename = `archive-${date.getFullYear()}-${date.getMonth() + 1}-${date.getDate()}.focalboard`

                    const file = new Blob([blob], {type: 'application/octet-stream'})
                    link.href = URL.createObjectURL(file)
                    link.download = filename
                    document.body.appendChild(link)						// FireFox support

                    link.click()

                    // TODO: Review if this is needed in the future, this is to fix the problem with linux webview links
                    if (window.openInNewBrowser) {
                        window.openInNewBrowser(link.href)
                    }

                    // TODO: Remove or reuse link and revolkObjectURL to avoid memory leak
                })
        })
    }

    private static async importBlocksFromFile(file: File): Promise<void> {
        let blockCount = 0
        const maxBlocksPerImport = 1000
        let blocks: Block[] = []

        let isFirstLine = true
        return new Promise<void>((resolve) => {
            LineReader.readFile(file, async (line, completed) => {
                if (completed) {
                    if (blocks.length > 0) {
                        await mutator.importFullArchive(blocks)
                        blockCount += blocks.length
                    }
                    Utils.log(`Imported ${blockCount} blocks.`)
                    resolve()
                    return
                }

                if (isFirstLine) {
                    isFirstLine = false
                    const header = JSON.parse(line) as ArchiveHeader
                    if (header.date && header.version >= 1) {
                        const date = new Date(header.date)
                        Utils.log(`Import archive, version: ${header.version}, date/time: ${date.toLocaleString()}.`)
                    }
                } else {
                    const row = JSON.parse(line) as ArchiveLine
                    if (!row || !row.type || !row.data) {
                        Utils.logError('importFullArchive ERROR parsing line')
                        return
                    }
                    switch (row.type) {
                    case 'block': {
                        const blockLine = row as BlockArchiveLine
                        const block = blockLine.data
                        if (Archiver.isValidBlock(block)) {
                            blocks.push(block)
                            if (blocks.length >= maxBlocksPerImport) {
                                const blocksToSend = blocks
                                blocks = []
                                await mutator.importFullArchive(blocksToSend)
                                blockCount += blocksToSend.length
                            }
                        }
                        break
                    }
                    }
                }
            })
        })
    }

    static isValidBlock(block: Block): boolean {
        if (!block.id || !block.rootId) {
            return false
        }

        return true
    }

    static importFullArchive(onComplete?: () => void): void {
        const input = document.createElement('input')
        input.type = 'file'
        input.accept = '.focalboard'
        input.onchange = async () => {
            const file = input.files && input.files[0]
            if (file) {
                await Archiver.importBlocksFromFile(file)
            }

            onComplete?.()
        }

        input.style.display = 'none'
        document.body.appendChild(input)
        input.click()

        // TODO: Remove or reuse input
    }
}

export {Archiver}
