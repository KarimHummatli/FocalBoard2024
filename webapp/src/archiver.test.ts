// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {fireEvent, waitFor} from '@testing-library/dom'
import {mocked} from 'ts-jest/utils'

import {TextDecoder} from 'web-encoding'

import {Utils} from './utils'
import {IAppWindow} from './types'
import {Archiver} from './archiver'
import {createBlock} from './blocks/block'
import mutator from './mutator'
import {TestBlockFactory} from './test/testBlockFactory'

jest.mock('./mutator')
jest.mock('./utils.ts')
const mockedMutator = mocked(mutator, true)
const mockedUtils = mocked(Utils, true)
const mockedIAppWindow = mocked(window as unknown as IAppWindow, true)

global.TextDecoder = TextDecoder

describe('src/archiver', () => {
    const board = TestBlockFactory.createBoard()
    board.id = 'board1'
    describe('exportBoardArchive', () => {
        beforeEach(() => {
            jest.useFakeTimers('modern')
            jest.setSystemTime(new Date('2021-11-05'))
            document.body.innerHTML = ''
            jest.clearAllMocks()
            mockedIAppWindow.openInNewBrowser = jest.fn()
        })
        afterEach(jest.useRealTimers)
        test('should export board archive', async () => {
            const mockedBlock = createBlock()
            mockedBlock.id = board.id
            mockedMutator.exportArchive.mockResolvedValue([mockedBlock])
            await Archiver.exportBoardArchive(board)
            expect(mockedMutator.exportArchive).toBeCalledWith(board.id)
            expect(mockedIAppWindow.openInNewBrowser).toBeCalled()
            expect(document.body).toMatchSnapshot()
        })
        test('should export full archive', async () => {
            const mockedBlock = createBlock()
            mockedBlock.id = board.id
            mockedMutator.exportArchive.mockResolvedValue([mockedBlock])
            await Archiver.exportFullArchive()
            expect(mockedMutator.exportArchive).toBeCalled()
            expect(mockedIAppWindow.openInNewBrowser).toBeCalled()
            expect(document.body).toMatchSnapshot()
        })
    })
    describe('isValidBlock', () => {
        beforeEach(() => {
            jest.clearAllMocks()
            document.body.innerHTML = ''
        })
        test('should validate block', () => {
            const mockedBlock = createBlock()
            mockedBlock.id = board.id
            mockedBlock.rootId = board.id
            expect(Archiver.isValidBlock(mockedBlock)).toBeTruthy()
        })
        test('should not validate block', () => {
            const mockedBlock = createBlock()
            expect(Archiver.isValidBlock(mockedBlock)).toBeFalsy()
        })
    })
    describe('importFullArchive', () => {
        const mockedComplete = jest.fn()
        beforeEach(() => {
            jest.clearAllMocks()
            document.body.innerHTML = ''
        })
        test('should import full archive', async () => {
            await Archiver.importFullArchive(mockedComplete)
            const elementInput = document.querySelector('input')
            fireEvent.change(elementInput!, {
                target: {
                    files: [new File(['{"version":1,"date":1636070400000}\n{"type":"block","data":{"id":"board1","schema":1,"workspaceId":"","parentId":"","rootId":"board1","createdBy":"","modifiedBy":"","type":"unknown","fields":{},"title":"","createAt":1636070400000,"updateAt":1636070400000,"deleteAt":0}}\n{"type":"block","data":{"id":"board2","schema":2,"workspaceId":"","parentId":"","rootId":"board2","createdBy":"","modifiedBy":"","type":"unknown","fields":{},"title":"","createAt":1636070400000,"updateAt":1636070400000,"deleteAt":0}}'], 'myfocalboard.focalboard', {type: 'text/json'})],
                },
            })
            await waitFor(async () => {
                expect(mockedComplete).toBeCalled()
                expect(mockedMutator.importFullArchive).toBeCalledTimes(1)
            })
        })
        test('shouldn\'t import full archive', async () => {
            await Archiver.importFullArchive(mockedComplete)
            const elementInput = document.querySelector('input')
            fireEvent.change(elementInput!, {
                target: {
                    files: [new File(['{"version":1,"date":1636070400000}\n{"type":"block"}'], 'myfocalboard.focalboard', {type: 'text/json'})],
                },
            })
            await waitFor(async () => {
                expect(mockedUtils.logError).toBeCalledWith('importFullArchive ERROR parsing line')
            })
        })
    })
})
