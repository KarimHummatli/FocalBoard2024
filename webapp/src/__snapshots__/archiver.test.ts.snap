// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/archiver exportBoardArchive should export board archive 1`] = `
<body>
  <a
    download="archive-2021-11-5.focalboard"
    href="data:text/json,%7B%22version%22%3A1%2C%22date%22%3A1636070400000%7D%0A%7B%22type%22%3A%22block%22%2C%22data%22%3A%7B%22id%22%3A%22board1%22%2C%22schema%22%3A1%2C%22workspaceId%22%3A%22%22%2C%22parentId%22%3A%22%22%2C%22rootId%22%3A%22%22%2C%22createdBy%22%3A%22%22%2C%22modifiedBy%22%3A%22%22%2C%22type%22%3A%22unknown%22%2C%22fields%22%3A%7B%7D%2C%22title%22%3A%22%22%2C%22createAt%22%3A1636070400000%2C%22updateAt%22%3A1636070400000%2C%22deleteAt%22%3A0%7D%7D%0A"
    style="display: none;"
  />
</body>
`;

exports[`src/archiver exportBoardArchive should export full archive 1`] = `
<body>
  <a
    download="archive-2021-11-5.focalboard"
    href="data:text/json,%7B%22version%22%3A1%2C%22date%22%3A1636070400000%7D%0A%7B%22type%22%3A%22block%22%2C%22data%22%3A%7B%22id%22%3A%22board1%22%2C%22schema%22%3A1%2C%22workspaceId%22%3A%22%22%2C%22parentId%22%3A%22%22%2C%22rootId%22%3A%22%22%2C%22createdBy%22%3A%22%22%2C%22modifiedBy%22%3A%22%22%2C%22type%22%3A%22unknown%22%2C%22fields%22%3A%7B%7D%2C%22title%22%3A%22%22%2C%22createAt%22%3A1636070400000%2C%22updateAt%22%3A1636070400000%2C%22deleteAt%22%3A0%7D%7D%0A"
    style="display: none;"
  />
</body>
`;
